### DISCUSSION TITLE
Use matrix in reusable workflows

### DISCUSSION THREAD

**Discussion by javixeneize** [community/community#29646] [Other Feature Feedback, Questions, & Ideas] [answered]
Hi

I am trying to use a matrix in a job where i am calling a reusable workflow. Looks like this is not supported by github

`Cannot define both `uses` and `strategy` at the same time for the following jobs: testjavi
`

Is there any workaround or any plan to support this? That would be a great addition

Thanks

**Comment by airtower-luna** [comment]
I'd try to pass the matrix to the reusable workflow via an input as a JSON string, and then use `fromJSON()` to set a job matrix from that JSON string inside the reusable workflow. I haven't tried that myself, but I've seen others use that approach successfully to create a matrix.

**Reply by javixeneize** [reply]
Thanks. I think you mean using fromJSON() under strategy, so it will not work either. The problem is to mix strategy and reusable workflows, even if you are not using a matrix underneath

Do you have any example where this is working fine? I have been unable to find any that fits my scenario

Thanks

**Reply by airtower-luna** [reply]
> I think you mean using fromJSON() under strategy, so it will not work either. The problem is to mix strategy and reusable workflows, even if you are not using a matrix underneath

My point is to set the matrix strategy on a job *inside* the reusable workflow, using `fromJSON()` on an input. That way you can pass the matrix data to the reusable workflow from the calling workflow, and avoid using `strategy` together with `uses`.

**Reply by javixeneize** [reply]
understood, yeah, but this will limit the reusability of the reusable workflow as sometimes i wont need a matrix, i want to control that from the wf calling the reusable workflow

**Reply by airtower-luna** [reply]
> this will limit the reusability of the reusable workflow as sometimes i wont need a matrix

You could provide a default for the input that results in just a single job.

And yes, I understand the difference, but you asked about possible workarounds. :slightly_smiling_face:

**Reply by javixeneize** [reply]
yeah that could work. Thanks!!

**Reply by javixeneize** [reply]
Hi

Is there any example on how to use a matrix in a reusable workflow with fromJson() in the reusable workflow?

I have followed this -> https://stackoverflow.com/questions/59977364/github-actions-how-use-strategy-matrix-with-script and it works fine in a single yaml with multiple jobs, as per the example. But when i put this in my reusable workflow i have this error:

"Invalid type found: object was expected but string was found"

This is what i have in the reusable workflow

jobs:
  build:
    runs-on: ubuntu-20.04
    strategy:
      matrix: ${{fromJson(inputs.matrix)}}

The input is a string, so this is clearly the problem, but, how can i pass this as an object?

Thanks

**Reply by airtower-luna** [reply]
The input being a string is fine, because `fromJSON()` needs a string to then parse as JSON. The question is, what kind of object comes out after parsing the input as JSON? Obviously I can't tell that without seeing the input value. :slightly_smiling_face:

**Reply by javixeneize** [reply]
it is a real json. The funny thing is the workflow works, but when editing the workflow in the browser directly, reports that error, which is not a real error!

**Reply by airtower-luna** [reply]
Oh, I thought that was an error from the workflow log! If the workflow works (including the matrix) that sounds like a case for a [bug report](https://support.github.com/contact/bug-report) about the web editor. :sweat_smile: 

> it is a real json.

Sure, I meant that depending on the JSON data the result of parsing it could be a string object. But if the workflow works that's probably not the case. :grin:

**Comment by airtower-luna** [comment, marked as answer]
Looks like matrix strategy for calling reusable workflows is supported now! :grin:  https://github.blog/changelog/2022-08-22-github-actions-improvements-to-reusable-workflows-2/

**Reply by javixeneize** [reply]
Wow!! Thats a huge improvement! Thanks 

I have another issue with outputs in matrix, https://github.com/orgs/community/discussions/30425 no idea if this upgrade would solve that problem too

**Reply by kaplanben** [reply]
+1 on the output issue... looks like reusable workflows with matrix support a single value for an output, meaning that you can't record (and later access) data from the matrixes. I couldn't find any alternative (besides recording it anywhere outside the runner).
Any chance this is a feature in planning?
